src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+                // should handle cases where call latencies are stable
+                if (scopeIterationTime === this.activeIterationTime) {
+                    this.quickPick.items = reset ? newItems.value : this.quickPick.items.concat(newItems.value)
+                    // nothing else to reset from here on out
+                    reset = false
+                    this.isDone = newItems.done ?? false
+                } else {
+                    // another newer iteration cycle is in-flight. Break without mutating state.
+                    break
+                }
+            } catch (e) {
+                // append error node
+                // clicking error node should either go backwards (return undefined) or refresh
+                // maybe have one of each?
+                // give quickpick an error message
+                // we should not blow away the existing items, they should still be viable
+                const err = e as Error
+                this.quickPick.items = [
+                    ...this.quickPick.items,
+                    {
+                        ...IteratingQuickPickController.ERROR_ITEM,
+                        detail: err.message,
+                    },
+                ]
+                this.isDone = true
+            }
+        }
+
+        // wrap up, but only if the fucntion call is the currently active one.
+        if (scopeIterationTime === this.activeIterationTime) {
+            // no items in response
+            if (this.isDone && this.quickPick.items.length === 0) {
+                this.quickPick.items = [IteratingQuickPickController.NO_ITEMS_ITEM]
+            }
+
+            // disable loading bar
+            this.quickPick.busy = false
+        }
+    }
+}
+
+/**
+ * Represents an iterator that tranforms another iterator into an array of QuickPickItems.
+ * Additionally, has a reset functionality to reset the iterator to its initial state.
+ */
+export class IteratingQuickPickPopulator<TResponse> {
+    private iterator: AsyncIterator<TResponse>
+
+    /**
+     * @param iteratorFactory Function that returns an iterator, with all default state values set. E.g. `collectionUtils.getPaginatedAwsCallIter`
+     * @param transform Function which transforms a response from the iterator into an array of `vscode.QuickPickItem`s.
+     */
+    public constructor(
+        private readonly iteratorFactory: () => AsyncIterator<TResponse>,
+        private readonly transform: (response: TResponse) => vscode.QuickPickItem[]
+    ) {
+        this.iterator = this.iteratorFactory()
+    }
+
+    /**
+     * Resets the iterator to the default state provided by the iteratorFactory.
+     */
+    public reset(): void {
+        this.iterator = this.iteratorFactory()
+    }
+
+    /**
+     * Generates an iterator which returns an array of formatted QuickPickItems on `.next()`
+     */
+    public async *getPickIterator(): AsyncIterator<vscode.QuickPickItem[]> {
```

Comment:
It's usually considered best practice to reserve the `get` and set` prefixes for functions that just return simple properties. Would use something like `create` or `build` here to signify that it's creating and returning a new iterator instance each time.